src/components/modals/ManagePlanDerivationGroupsModal.svelte

```diff
@@ -0,0 +1,362 @@
+  }>();
+
+  let externalEventLayers: Layer[] | undefined;
+  let dataGrid: DataGrid<DerivationGroup>;
+  let baseColumnDefs: DataGridColumnDef<DerivationGroup>[] = [];
+
+  selectedPlanDerivationGroupNames.subscribe(() => {
+    if (baseColumnDefs.length > 0 && dataGrid) {
+      // no current way to change just a specific cell unless we add something about plan associations to the DG object,
+      //    which we don't seek to do.
+      // this does mean every update to any entry in selectedPlanDerivationGroupIds refreshes the whole column, and flashes it,
+      //    though that isn't particularly a problem and does do good to signal association complete. Also a small delay, which
+      //    buffers button smashing and repeated updates pretty well!
+      dataGrid.refreshCells();
+    }
+  });
+
+  $: externalEventLayers = $view?.definition.plan.timelines
+    .flatMap(timeline => timeline.rows)
+    .flatMap(row => row.layers)
+    .filter(layer => isExternalEventLayer(layer));
+
+  baseColumnDefs = [
+    {
+      field: 'name',
+      filter: 'string',
+      headerName: 'Derivation Group',
+      resizable: true,
+      sortable: true,
+      suppressAutoSize: false,
+      suppressSizeToFit: false,
+    },
+    {
+      field: 'source_type_name',
+      filter: 'string',
+      headerName: 'Source type',
+      resizable: true,
+      sortable: true,
+      suppressAutoSize: false,
+      suppressSizeToFit: false,
+    },
+    {
+      field: 'derived_event_total',
+      filter: 'number',
+      headerName: 'Derived Events in Derivation Group',
+      sortable: true,
+      suppressAutoSize: true,
+      suppressSizeToFit: true,
+      valueFormatter: params => {
+        return params?.value.length;
+      },
+    },
+    {
+      cellRenderer: (params: DerivationGroupCellRendererParams) => {
+        var input = document.createElement('input');
+        input.type = 'checkbox';
+        input.checked = $selectedPlanDerivationGroupNames.includes(params?.data?.name ?? '');
+        input.addEventListener('click', event => {
+          if (event?.target && params.data) {
+            if ((event.target as any).checked) {
+              // insert
+              effects.insertDerivationGroupForPlan(params.data.name, $plan, user);
+              if ($derivationGroupPlanLinkError !== null) {
+                showFailureToast('Failed to link derivation group & plan.');
+              } else {
+                // Insert all the external event types from the derivation group to the timeline filter
+                const derivationGroup = $derivationGroups.find(
+                  derivationGroup => derivationGroup.name === params?.data?.name,
+                );
+                if (derivationGroup !== undefined) {
+                  externalEventLayers?.forEach(externalEventLayer => {
+                    if (externalEventLayer.filter.externalEvent !== undefined) {
+                      externalEventLayer.filter.externalEvent.event_types =
```

Comment:
How else, besides mutation, should we update the event_types? Are you asking if we can use a store or something else entirely?

Reply:
The problem with mutations in svelte is that if there are components that rely on `externalEventLayer.filter`, for instance, those components will not get triggered to rerender when `event_types` gets replaced. The way we would update something like this is to completely remake the object so that all the pointers are explicitly changed.

So rather than doing something like
```
externalEventLayers.forEach(...)
```
we do something like
```
externalEventLayers = externalEventLayers.map(layer => {
  return {
    ...layer,
    filter: {
      ...layer.filter,
      externalEvent: {
        ...layer.filter.externalEvent,
        event_types: (...)
      }
    }
  }
});
```

Reply:
See [this commit](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/5756e9fe9c0f49dd6f0454df66f03882053a7f0a).

Reply:
Piggy-backing this to add the following commit: [ Update all related external event layers upon derivation group association](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/618f745faa194a2014b436e6550d48e986416568)

Pranav's commit does the work of mapping new layers, and then mine above does the work of applying those over the old layers. I used the handler logic from how activity type filter changes are applied from the timeline editor in order to construct this, though it was a bit more difficult as associating derivation groups with a plan is, by design, row/layer/timeline agnostic. It seems to function correctly after this change, but let me know if this fits what you had in mind cause I was definitely having a little trouble making sure I was doing this correctly! Additionally, I think we'll want to replicate this for disassociating derivation groups as well because in that case we may need to 'regenerate' the Layer to remove the event type from the filter.